---
title: "PCA_permanovas"
author: "allyson_demerlis"
date: "2022-12-30"
output: html_document
---
# Install and load necessary packages
```{r, install/load packages}
#if you need to install all the packages, you need the package "BiocManager" first.
# if (!require("BiocManager",quietly=TRUE))
#   install.packages("BiocManager")
#Now you can use BiocManager to install pacakges like DESeq2
#BiocManager::install("DESeq2")

library(DESeq2)

library(DEGreport) # needs BiocManager
  #DEGreport requires lasso2, which is not available for newer versions of R. To install lasso2, run library(remotes) (might need to install that package first), then run:
#install_version("lasso2", "1.2-22")

library(tidyverse) 
library(EnhancedVolcano) # needs BiocManager
library(ggpubr)
library(apeglm) # needs BiocManager
library(topGO) # needs BiocManager
library(GO.db) # needs BiocManager
library(Rgraphviz) # needs BiocManager
#library(factoextra) #for PCA and eigenvectors
#library(vegan) #for adonis function for PCA
library(VennDiagram)

library(scales) #for GO dot plot
library(viridis) #for GO dot plot
```


# Importing gene counts matrix and formatting it
```{r}
#reading in txt file as a table with the headers as the header in the original file
countsmatrix <-read.table("../WH_Pdam.Rmatrix.txt",header=TRUE)

#removing the Chromosome column from the dataset, and changing the rows to be the gene names
countsmatrix <-countsmatrix %>% 
  dplyr::select(-c("Chr")) %>% 
  column_to_rownames("Geneid")

#removing hour 5 from the dataset because it only has 1 wounded coral, so it is not comparable 
countsmatrix <- countsmatrix %>% 
  dplyr::select(!X5331C1:X5335Y)

#function found from stack overflow
#tidying column names so they don't include an X anymore (R puts an X in front of columns that start with numbers)
destroyX = function(es) {
  f = es
  for (col in c(1:ncol(f))){ #for each column in dataframe
    if (startsWith(colnames(f)[col], "X") == TRUE)  { #if starts with 'X' ..
      colnames(f)[col] <- substr(colnames(f)[col], 2, 100) #get rid of it
    }
  }
  assign(deparse(substitute(es)), f, inherits = TRUE) #assign corrected data to original name
}

destroyX(countsmatrix)
```


# Import gene feature annotation information
```{r}
genefeatures <- read.delim(file = "../input files/pdam_genome_IDInfo.gff", header = F)
head(genefeatures)
colnames(genefeatures) <- c("IDGeneInfo")
rownames(genefeatures) <- rownames(countsmatrix)

#removing gene name from the second column
gene_name_split <- str_split_fixed(genefeatures$IDGeneInfo, " ", 2)
print(head(gene_name_split))
colnames(gene_name_split)<- c("Gene_ID","Gene_Function")
head(gene_name_split)
gene_name_split_df<-data.frame(gene_name_split)
head(gene_name_split_df)

#convert the Gene_ID column to the row headers
gene_name_split_df %>%
     remove_rownames() %>%
 column_to_rownames(var = 'Gene_ID')-> Pdam_Gene_Names_Info 
#%>%
  #write.csv(file="Pdam_Gene_Names_Info.csv")
```

# Make metadata file, DDS object, and DESeq2 results for each hour

We are splitting up the hours to compare wounded versus control conditions at each time point. DESeq2 compares differential gene expression of two groups using the Wald test, so we need to subset our data to do each comparison directly.

## Hour zero
```{r}
countdata_0 <- countsmatrix %>% dplyr::select(`0301C1`:`0306Z`)

metadata_0 = data.frame(sample=colnames(countdata_0),
                condition = stringr::str_detect(pattern = ".*C.*",string = colnames(countdata_0)),
                hour = stringr::str_replace(pattern = "(.).*",replacement="\\1",string = colnames(countdata_0)),
 id = stringr::str_replace(pattern=".(...).*",replacement="\\1",string=colnames(countdata_0)))

#changing TRUE and FALSE for condition to Control and Wounded
metadata_0$condition[str_detect(metadata_0$condition,"TRUE")] <- "Control"
metadata_0$condition[str_detect(metadata_0$condition,"FALSE")] <- "Wounded"

metadata_0 <- metadata_0 %>% column_to_rownames("sample")

metadata_0$condition <- as.factor(metadata_0$condition)

#DESeq2 from a counts matrix; requires count data, metadata, and the experimental design
dds_0=DESeq2::DESeqDataSetFromMatrix(countData = countdata_0, colData = metadata_0, design = ~condition)
dds_0 <- estimateSizeFactors(dds_0)
dds_0$sizeFactor #this is based on an internal normalization where geometric mean is calculated for each gene across all samples, then the counts for a gene in each sample is then divided by this mean. The median of these ratios in a sample is the size factor for that sample (https://rdrr.io › Bioconductor › DESeq2)

#prefiltering recommended by DESeq2 Vignette (removes anything with reads below 10)
keep_0 <- rowSums(counts(dds_0)) >= 10
length(keep_0) #26077
dds_0 <- dds_0[keep_0,]
length(dds_0) #19447

dds_0<-DESeq(dds_0)

res_h0<-results(dds_0)
summary(res_h0,alpha=0.05)
#5 downregulated genes (3 outliers), no upregulated genes

#identifying significant genes
resSig_h0<-res_h0[which(res_h0$padj<0.05), ]

#annotate results to include Gene Function
anno_h0<-merge(as.data.frame(resSig_h0),Pdam_Gene_Names_Info,by='row.names',all=TRUE)
anno_h0<-na.omit(anno_h0)
```

## Hour 1
```{r}
countdata_1 <- countsmatrix %>% dplyr::select(`1307C1`:`1312Z`)

metadata_1 = data.frame(sample=colnames(countdata_1),
                condition = stringr::str_detect(pattern = ".*C.*",string = colnames(countdata_1)),
                hour = stringr::str_replace(pattern = "(.).*",replacement="\\1",string = colnames(countdata_1)),
                id = stringr::str_replace(pattern=".(...).*",replacement="\\1",string=colnames(countdata_1)))

#changing TRUE and FALSE for condition to Control and Wounded
metadata_1$condition[str_detect(metadata_1$condition,"TRUE")] <- "Control"
metadata_1$condition[str_detect(metadata_1$condition,"FALSE")] <- "Wounded"

metadata_1 <- metadata_1 %>% column_to_rownames("sample")

metadata_1$condition <- as.factor(metadata_1$condition)

#DESeq2 from a counts matrix; requires count data, metadata, and the experimental design
dds_1=DESeq2::DESeqDataSetFromMatrix(countData = countdata_1, colData = metadata_1, design = ~condition)
dds_1 <- estimateSizeFactors(dds_1)

#prefiltering recommended by DESeq2 Vignette (removes anything with reads below 10)
keep_1 <- rowSums(counts(dds_1)) >= 10
length(keep_1) #26077
dds_1 <- dds_1[keep_1,]
length(dds_1) #19536

dds_1<-DESeq(dds_1)

res_h1<-results(dds_1)
summary(res_h1,alpha=0.05)
#26 upregulated genes, 49 downregulated, 1 outlier, 758 low counts

#identifying significant genes
resSig_h1<-res_h1[which(res_h1$padj<0.05), ]

#annotate results to include Gene Function
anno_h1<-merge(as.data.frame(resSig_h1),Pdam_Gene_Names_Info,by='row.names',all=TRUE)
anno_h1<-na.omit(anno_h1)
```

## Hour 2
```{r}
countdata_2 <- countsmatrix %>% dplyr::select(`2313C1`:`2318Z`)

metadata_2 = data.frame(sample=colnames(countdata_2),
                condition = stringr::str_detect(pattern = ".*C.*",string = colnames(countdata_2)),
                hour = stringr::str_replace(pattern = "(.).*",replacement="\\1",string = colnames(countdata_2)),
                id = stringr::str_replace(pattern=".(...).*",replacement="\\1",string=colnames(countdata_2)))

#changing TRUE and FALSE for condition to Control and Wounded
metadata_2$condition[str_detect(metadata_2$condition,"TRUE")] <- "Control"
metadata_2$condition[str_detect(metadata_2$condition,"FALSE")] <- "Wounded"

metadata_2 <- metadata_2 %>% column_to_rownames("sample")

metadata_2$condition <- as.factor(metadata_2$condition)

#DESeq2 from a counts matrix; requires count data, metadata, and the experimental design
dds_2=DESeq2::DESeqDataSetFromMatrix(countData = countdata_2, colData = metadata_2, design = ~condition)
dds_2 <- estimateSizeFactors(dds_2)

#prefiltering recommended by DESeq2 Vignette (removes anything with reads below 10)
keep_2 <- rowSums(counts(dds_2)) >= 10
length(keep_2) #26077
dds_2 <- dds_2[keep_2,]
length(dds_2) #19779

dds_2<-DESeq(dds_2)

res_h2<-results(dds_2)
summary(res_h2,alpha=0.05)
#6 upregulated genes, 10 downregulated, 8 outliers, 6517 low counts

#identifying significant genes
resSig_h2<-res_h2[which(res_h2$padj<0.05), ]

#annotate results to include Gene Function
anno_h2<-merge(as.data.frame(resSig_h2),Pdam_Gene_Names_Info,by='row.names',all=TRUE)
anno_h2<-na.omit(anno_h2)
```

## Hour 4
```{r}
countdata_4 <- countsmatrix %>% dplyr::select(`4325C1`:`4330Z`)

metadata_4 = data.frame(sample=colnames(countdata_4),
                condition = stringr::str_detect(pattern = ".*C.*",string = colnames(countdata_4)),
                hour = stringr::str_replace(pattern = "(.).*",replacement="\\1",string = colnames(countdata_4)),
                id = stringr::str_replace(pattern=".(...).*",replacement="\\1",string=colnames(countdata_4)))

#changing TRUE and FALSE for condition to Control and Wounded
metadata_4$condition[str_detect(metadata_4$condition,"TRUE")] <- "Control"
metadata_4$condition[str_detect(metadata_4$condition,"FALSE")] <- "Wounded"

metadata_4 <- metadata_4 %>% column_to_rownames("sample")

metadata_4$condition <- as.factor(metadata_4$condition)

#DESeq2 from a counts matrix; requires count data, metadata, and the experimental design
dds_4=DESeq2::DESeqDataSetFromMatrix(countData = countdata_4, colData = metadata_4, design = ~condition)
dds_4 <- estimateSizeFactors(dds_4)

#prefiltering recommended by DESeq2 Vignette (removes anything with reads below 10)
keep_4 <- rowSums(counts(dds_4)) >= 10
length(keep_4) #26077
dds_4 <- dds_4[keep_4,]
length(dds_4) #19773

dds_4<-DESeq(dds_4)

res_h4<-results(dds_4)
summary(res_h4,alpha=0.05)
#19 upregulated, 27 downregulated, 31 outliers, 1151 low counts

#identifying significant genes
resSig_h4<-res_h4[which(res_h4$padj<0.05), ]

#annotate results to include Gene Function
anno_h4<-merge(as.data.frame(resSig_h4),Pdam_Gene_Names_Info,by='row.names',all=TRUE)
anno_h4<-na.omit(anno_h4)
```

# Combining all annotated differentially expressed genes (padj<0.05) into a table
```{r}
anno_h0 %>% 
  mutate(time = "hour 0") -> anno_h0

anno_h1 %>% 
  mutate(time = "hour 1") -> anno_h1

anno_h2 %>% 
  mutate(time = "hour 2") -> anno_h2

anno_h4 %>% 
  mutate(time = "hour 4") -> anno_h4

full_join(anno_h0, anno_h1) -> annotated_DEG_list
full_join(anno_h2, annotated_DEG_list) -> annotated_DEG_list
full_join(anno_h4, annotated_DEG_list) -> annotated_DEG_list

#write_csv(annotated_DEG_list, "../2022 updates with Sami Beasley/annotated_DEG_list_allhours.csv")
```

# Principal component analyses and Scree plots
```{r}
#need to transform the data for plotting
dds_vst0<- vst(dds_0,blind=FALSE)
plotPCA(dds_vst0)
#plotPCA is one function to do it, or you can use the "prcomp" function, which lets you create a scree plot
pca_h0 <- prcomp(t(assay(dds_vst0)))
fviz_eig(pca_h0)

#you can also manually create a scree plot using the code below
# percentVar_h0 <-pca_h0$sdev^2 / sum( pca_h0$sdev^2 ) #the contribution to the total variance for each component
# scree_plot_h0=data.frame(percentVar_h0)
# scree_plot_h0[,2]<- c(1:6)
# colnames(scree_plot_h0)<-c("variance","component_number")
# ggplot(scree_plot_h0, mapping=aes(x=component_number, y=variance))+geom_bar(stat="identity")

dds_vst1<- vst(dds_1,blind=FALSE)
plotPCA(dds_vst1)
pca_h1 <- prcomp(t(assay(dds_vst1)))
fviz_eig(pca_h1)

dds_vst2<- vst(dds_2,blind=FALSE)
plotPCA(dds_vst2)
pca_h2 <- prcomp(t(assay(dds_vst2)))
fviz_eig(pca_h2)

dds_vst4<- vst(dds_4,blind=FALSE)
plotPCA(dds_vst4)
pca_h4 <- prcomp(t(assay(dds_vst4)))
fviz_eig(pca_h4)
```


## PCA figures for manuscript
```{r}
#plotting the PCA in ggplot
pca12_0 <- plotPCA(dds_vst0,intgroup=c("condition"),returnData = TRUE)
pca_0 = ggplot(pca12_0, aes(PC1,PC2,shape=condition,color=condition)) + 
  geom_point(size=3) +  
  xlab(paste0("PC1 (49%)")) +
  ylab(paste0("PC2 (24%)")) +
  theme(legend.position="right")  + 
  theme(text = element_text(size=12))  + 
  theme(legend.key.size = unit(0.5, "cm")) + 
  theme(legend.title=element_text(size=12)) +
  scale_shape_discrete(name="Condition") +
  scale_color_discrete(name="Condition") +
  theme_classic(base_size=12,base_family="serif")

pca12_1 <- plotPCA(dds_vst1,intgroup=c("condition"),returnData = TRUE)
pca_1 = ggplot(pca12_1, aes(PC1, PC2,shape=condition,color=condition)) + 
  geom_point(size=3) + 
  xlab(paste0("PC1 (38%)")) +
  ylab(paste0("PC2 (23%)")) +
  theme(legend.position="right")  + 
  theme(text = element_text(size=12)) + 
  theme(legend.key.size = unit(0.5, "cm")) + 
  theme(legend.title=element_text(size=12)) +
  scale_shape_discrete(name="Condition") +
  scale_color_discrete(name="Condition") +
  theme_classic(base_size=12,base_family="serif")

pca12_2 <- plotPCA(dds_vst2,intgroup=c("condition"),returnData = TRUE)
pca_2 = ggplot(pca12_2, aes(PC1, PC2,shape=condition,color=condition)) + 
  geom_point(size=3) +   
  xlab(paste0("PC1 (35%)")) +
  ylab(paste0("PC2 (29%)")) +
  theme(legend.position="right")  + 
  theme(text = element_text(size=12)) + 
  theme(legend.key.size = unit(0.5, "cm")) + 
  theme(legend.title=element_text(size=12)) +
  scale_shape_discrete(name="Condition") +
  scale_color_discrete(name="Condition") +
  theme_classic(base_size=12,base_family="serif")

pca12_4 <- plotPCA(dds_vst4,intgroup=c("condition"),returnData = TRUE)
pca_4 = ggplot(pca12_4, aes(PC1, PC2,shape=condition,color=condition)) + 
  geom_point(size=3) +  
  xlab(paste0("PC1 (56%)")) +
  ylab(paste0("PC2 (21%)")) +
  theme(legend.position="right")  + 
  theme(text = element_text(size=12))  + 
  theme(legend.key.size = unit(0.5, "cm")) + 
  theme(legend.title=element_text(size=12)) +
  scale_shape_discrete(name="Condition") +
  scale_color_discrete(name="Condition") +
  theme_classic(base_size=12,base_family="serif")

pca_combined<-ggarrange(pca_0,pca_1,pca_2,pca_4,
          common.legend = TRUE,
          legend="right",
          labels=c("A","B","C","D"),
          ncol=2,nrow=2)

pca_combined
#brought this figure into Illustrator to edit colors and formatting
```



# PERMANOVAs
## for hour 0
```{r}
#we use the dds object instead of the original countdata_0 matrix because this has filtered out genes with low counts (<10)
hour0_countsmatrix <- assay(dds_0)
t(hour0_countsmatrix) -> hour0_countsmatrix 
PCA.h0.countsdata <- merge(metadata_0, hour0_countsmatrix, by='row.names', all=TRUE)

# scale data
vegan <- scale(PCA.h0.countsdata[c(5:19451)]) #we just want to scale the gene counts

# PERMANOVA 
permanova<-adonis2(vegan ~ condition, data = PCA.h0.countsdata, method='eu', na.rm=TRUE, nperm = 999)
permanova
```

## PERMANOVA for hour 1
```{r}
#we use the dds object instead of the original countdata_0 matrix because this has filtered out genes with low counts (<10)
hour1_countsmatrix <- assay(dds_1)
t(hour1_countsmatrix) -> hour1_countsmatrix 
PCA.h1.countsdata <- merge(metadata_1, hour1_countsmatrix, by='row.names', all=TRUE)

# scale data
vegan <- scale(PCA.h1.countsdata[c(5:19536)]) #we just want to scale the gene counts

# PERMANOVA 
permanova<-adonis2(vegan ~ condition, data = PCA.h1.countsdata, method='eu', na.rm=TRUE, nperm = 999)
permanova
```

## PERMANOVA for hour 2
```{r}
#we use the dds object instead of the original countdata_0 matrix because this has filtered out genes with low counts (<10)
hour2_countsmatrix <- assay(dds_2)
t(hour2_countsmatrix) -> hour2_countsmatrix 
PCA.h2.countsdata <- merge(metadata_2, hour2_countsmatrix, by='row.names', all=TRUE)

# scale data
vegan <- scale(PCA.h2.countsdata[c(5:19779)]) #we just want to scale the gene counts

# PERMANOVA 
permanova<-adonis2(vegan ~ condition, data = PCA.h2.countsdata, method='eu', na.rm=TRUE, nperm = 999)
permanova
```

## PERMANOVA for hour 4
```{r}
#we use the dds object instead of the original countdata_0 matrix because this has filtered out genes with low counts (<10)
hour4_countsmatrix <- assay(dds_4)
t(hour4_countsmatrix) -> hour4_countsmatrix 
PCA.h4.countsdata <- merge(metadata_4, hour4_countsmatrix, by='row.names', all=TRUE)

# scale data
vegan <- scale(PCA.h4.countsdata[c(5:19773)]) #we just want to scale the gene counts

# PERMANOVA 
permanova<-adonis2(vegan ~ condition, data = PCA.h4.countsdata, method='eu', na.rm=TRUE, nperm = 999)
permanova
```