---
title: "GO_analysis"
author: "allyson_demerlis"
date: "2022-10-25"
output: html_document
---

This markdown is for the Gene Ontology (GO) enrichment analysis using the package "topGO" (Alexa et al. 2006). 

# Install and load necessary packages
```{r, install/load packages}
#if you need to install all the packages, you need the package "BiocManager" first.
# if (!require("BiocManager",quietly=TRUE))
#   install.packages("BiocManager")
#Now you can use BiocManager to install pacakges like DESeq2
#BiocManager::install("DESeq2")

library(tidyverse) 
library(topGO) # needs BiocManager
library(GO.db) # needs BiocManager
library(Rgraphviz) # needs BiocManager
library(scales) #for GO dot plot
library(viridis) #for GO dot plot
```

# Import DDS objects for hours
```{r}
dds_1 <- readRDS("dds_1.rds")
res_h1 <- results(dds_1)

dds_2 <- readRDS("dds_2.rds")
res_h2 <- results(dds_2)

dds_4 <- readRDS("dds_4.rds")
res_h4 <- results(dds_4)
```


# topGO analysis (Molecular Function)
```{r}
#following Michael Connelly's code (github)

#Input required *P. damicornis* GO annotation data and construct Gene-to-GO object for custom annotation mapping
GO_geneID<-readMappings(file="../input_files//pdam_genome_GOgenes.txt") #Mike made this file

geneID_GO <- inverseList(GO_geneID)
str(head(geneID_GO))

#Generate gene universe and GO universe from Gene-to-GO and GO-to-Gene objects
geneNames <- names(geneID_GO)
str(head(geneNames))

GONames <- names(GO_geneID)
str(head(GONames))
```

Hour 0 had no annotations so no GO analysis

# Hour 1

## topGO analysis Hour 1 Molecular Function (MF)
```{r}
#label differentially expressed genes as '1' and non-significant genes as '0'
res_h1_GO<-as.data.frame(res_h1)
res_h1_GO<- res_h1_GO%>%mutate(Group=case_when(padj<0.05~"1",padj>=0.05~"0"))
res_h1_GO<-na.omit(res_h1_GO)

#create the gene universe
gene_universe_h1<-as.numeric(res_h1_GO$Group)
gene_universe_h1<-factor(gene_universe_h1)
names(gene_universe_h1) <- rownames(res_h1_GO) 

#create topGO data object
GO_MF_h1 <-new("topGOdata",
                ontology="MF",
                allGenes=gene_universe_h1,
                nodeSize=10,
                annotationFun=annFUN.gene2GO,
                gene2GO = geneID_GO)

# tests to identify enriched gene ontology terms
GO_MF_h1_resultFisher01<-runTest(GO_MF_h1,algorithm = "weight01", statistic = "fisher")
GO_MF_h1_resultFisherclassic<-runTest(GO_MF_h1,algorithm = "classic", statistic = "fisher")
GO_MF_h1_resultclassicKS <- runTest(GO_MF_h1, algorithm = "classic", statistic = "ks")
GO_MF_h1_resultelimKS <- runTest(GO_MF_h1, algorithm = "elim", statistic = "ks")


#generate results table
GO_MF_h1_resultstable<-GenTable(GO_MF_h1, 
                                Fisher01 = GO_MF_h1_resultFisher01,
                                Fisherclassic = GO_MF_h1_resultFisherclassic,
                                classicKS = GO_MF_h1_resultclassicKS,
                                elimKS = GO_MF_h1_resultelimKS,
                                topNodes = 50)

showSigOfNodes(GO_MF_h1,score(GO_MF_h1_resultelimKS),firstSigNodes = 5,useInfo = 'all')

GO_MF_h1_resultstable %>% filter(Significant >= 1) %>% mutate(hour=1)-> sigGO_MF_h1 #filter for genes from this time point that are significantly associated with the GO terms (which are also significantly enriched)
```

## splitting up- and down-regulated genes based on log2fold change for hour 1

### Up-regulated genes in hour 1
```{r}
#label differentially expressed genes as '1' and non-significant genes as '0'
res_GO_h1_up <- as.data.frame(res_h1) %>%
  mutate(Group=case_when(padj<0.05 & log2FoldChange >= 2 ~"1", padj>=0.05 ~"0")) #selecting log2fold change level of 2 because that matches volcano plots
res_GO_h1_up <- na.omit(res_GO_h1_up)

res_GO_h1_up %>% filter(Group == "1") #5 genes - matches volcano plot

#create the gene universe
gene_universe_h1_up <-as.numeric(res_GO_h1_up$Group)
gene_universe_h1_up <-factor(gene_universe_h1_up)
names(gene_universe_h1_up) <- rownames(res_GO_h1_up) 

#create topGO data object
GO_MF_h1_up <-new("topGOdata",
                ontology="MF",
                allGenes=gene_universe_h1_up,
                nodeSize=10,
                annotationFun=annFUN.gene2GO,
                gene2GO = geneID_GO)

# tests to identify enriched gene ontology terms
GO_MF_h1_up_resultFisher01<-runTest(GO_MF_h1_up,algorithm = "weight01", statistic = "fisher")
GO_MF_h1_up_resultFisherclassic<-runTest(GO_MF_h1_up,algorithm = "classic", statistic = "fisher")
GO_MF_h1_up_resultclassicKS <- runTest(GO_MF_h1_up, algorithm = "classic", statistic = "ks")
GO_MF_h1_up_resultelimKS <- runTest(GO_MF_h1_up, algorithm = "elim", statistic = "ks")


#generate results table
GO_MF_h1_up_resultstable<-GenTable(GO_MF_h1_up, 
                                Fisher01 = GO_MF_h1_up_resultFisher01,
                                Fisherclassic = GO_MF_h1_up_resultFisherclassic,
                                classicKS = GO_MF_h1_up_resultclassicKS,
                                elimKS = GO_MF_h1_up_resultelimKS,
                                topNodes = 50)

showSigOfNodes(GO_MF_h1_up,score(GO_MF_h1_up_resultelimKS),firstSigNodes = 5,useInfo = 'all')

GO_MF_h1_up_resultstable %>% 
  filter(Significant >= 1) %>% 
  mutate(hour=1)-> sigGO_MF_h1_up #filter for genes from this time point that are significantly associated with the GO terms (which are also significantly enriched)
```


## topGO analysis Hour 2 Molecular Function (MF)
```{r}
#label differentially expressed genes as '1' and non-significant genes as '0'
res_h2_GO<-as.data.frame(res_h2)
res_h2_GO<- res_h2_GO%>%mutate(Group=case_when(padj<0.05~"1",padj>=0.05~"0"))
res_h2_GO<-na.omit(res_h2_GO)

#create the gene universe
gene_universe_h2<-as.numeric(res_h2_GO$Group)
gene_universe_h2<-factor(gene_universe_h2)
names(gene_universe_h2) <- rownames(res_h2_GO) 

#create topGO data object
GO_MF_h2 <-new("topGOdata",
                ontology="MF",
                allGenes=gene_universe_h2,
                nodeSize=10,
                annotationFun=annFUN.gene2GO,
                gene2GO = geneID_GO)

# tests to identify enriched gene ontology terms
GO_MF_h2_resultFisher01<-runTest(GO_MF_h2,algorithm = "weight01", statistic = "fisher")
GO_MF_h2_resultFisherclassic<-runTest(GO_MF_h2,algorithm = "classic", statistic = "fisher")
GO_MF_h2_resultclassicKS <- runTest(GO_MF_h2, algorithm = "classic", statistic = "ks")
GO_MF_h2_resultelimKS <- runTest(GO_MF_h2, algorithm = "elim", statistic = "ks")


#generate results table
GO_MF_h2_resultstable<-GenTable(GO_MF_h2, 
                                Fisher01 = GO_MF_h2_resultFisher01,
                                Fisherclassic = GO_MF_h2_resultFisherclassic,
                                classicKS = GO_MF_h2_resultclassicKS,
                                elimKS = GO_MF_h2_resultelimKS,
                                topNodes = 50)

showSigOfNodes(GO_MF_h2,score(GO_MF_h2_resultelimKS),firstSigNodes = 5,useInfo = 'all')

GO_MF_h2_resultstable %>% filter(Significant >= 1) %>% mutate(hour=2)-> sigGO_MF_h2 #filter for genes from this time point that are significantly associated with the GO terms (which are also significantly enriched)
```

## topGO analysis Hour 4 Molecular Function (MF)
```{r}
#label differentially expressed genes as '1' and non-significant genes as '0'
res_h4_GO<-as.data.frame(res_h4)
res_h4_GO<- res_h4_GO%>%mutate(Group=case_when(padj<0.05~"1",padj>=0.05~"0"))
res_h4_GO<-na.omit(res_h4_GO)

#create the gene universe
gene_universe_h4<-as.numeric(res_h4_GO$Group)
gene_universe_h4<-factor(gene_universe_h4)
names(gene_universe_h4) <- rownames(res_h4_GO) 

#create topGO data object
GO_MF_h4 <-new("topGOdata",
                ontology="MF",
                allGenes=gene_universe_h4,
                nodeSize=10,
                annotationFun=annFUN.gene2GO,
                gene2GO = geneID_GO)

# tests to identify enriched gene ontology terms
GO_MF_h4_resultFisher01<-runTest(GO_MF_h4,algorithm = "weight01", statistic = "fisher")
GO_MF_h4_resultFisherclassic<-runTest(GO_MF_h4,algorithm = "classic", statistic = "fisher")
GO_MF_h4_resultclassicKS <- runTest(GO_MF_h4, algorithm = "classic", statistic = "ks")
GO_MF_h4_resultelimKS <- runTest(GO_MF_h4, algorithm = "elim", statistic = "ks")


#generate results table
GO_MF_h4_resultstable<-GenTable(GO_MF_h4, 
                                Fisher01 = GO_MF_h4_resultFisher01,
                                Fisherclassic = GO_MF_h4_resultFisherclassic,
                                classicKS = GO_MF_h4_resultclassicKS,
                                elimKS = GO_MF_h4_resultelimKS,
                                topNodes = 50)

showSigOfNodes(GO_MF_h4,score(GO_MF_h4_resultelimKS),firstSigNodes = 5,useInfo = 'all')

GO_MF_h4_resultstable %>% filter(Significant >= 1) %>% mutate(hour=4)-> sigGO_MF_h4 #filter for genes from this time point that are significantly associated with the GO terms (which are also significantly enriched)
```


## combine all significant gene GO lists per hour and visualize as table
```{r}
full_join(sigGO_MF_h1, sigGO_MF_h2) %>% full_join(., sigGO_MF_h4) %>%  #144 GO terms total
  filter(Fisherclassic < 0.05) #25 GO terms now
  #write_csv("GO_MF_allhours.csv")

#dot plot
full_join(sigGO_MF_h1, sigGO_MF_h2) %>% full_join(., sigGO_MF_h4) %>%
  filter(Fisherclassic < 0.05) %>% 
  mutate(Fisherclassic = as.numeric(Fisherclassic)) %>% 
  mutate(Annotated = as.numeric(Annotated)) %>% 
ggplot(data=., aes(x=-log(Fisherclassic), y=reorder(Term, -Fisherclassic))) +
    geom_point(aes(size=Annotated, color=-log(Fisherclassic))) +
    scale_color_viridis(option="plasma") +
  labs(x="-log P-Value", y="Gene Ontology Term Description", title=paste("Significant Molecular Functions"), color = "-log(P-value)") +
  scale_y_discrete(labels = function(x) str_wrap(x, width = 30)) +
  theme_bw() +
  facet_grid(vars(hour), scales = "free") + 
  scale_size(range = c(2,6))
##ggsave("significantMF_GO_perhour.pdf")

#need to make one that has the GO term next to it so I can easily google it for the illustrator manual edits
full_join(sigGO_MF_h1, sigGO_MF_h2) %>% full_join(., sigGO_MF_h4) %>%
  filter(Fisherclassic < 0.05) %>% 
  mutate(Fisherclassic = as.numeric(Fisherclassic)) %>% 
  mutate(Annotated = as.numeric(Annotated)) %>% 
  mutate(GO_term=str_c(GO.ID,Term, sep=" ")) %>% 
  ggplot(data=., aes(x=-log(Fisherclassic), y=reorder(GO_term, -Fisherclassic))) +
    geom_point(aes(size=Annotated, color=-log(Fisherclassic))) +
    scale_color_viridis(option="plasma") +
  labs(x="-log P-Value", y="Gene Ontology Term Description", title=paste("Significant Molecular Functions"), color = "-log(P-value)") +
  scale_y_discrete(labels = function(x) str_wrap(x, width = 50)) +
  theme_bw() +
  facet_grid(vars(hour), scales = "free") + 
  scale_size(range = c(2,6))
```



# Venn Diagrams
## for DGEs at each hour
```{r}
# need to manually calculate the totals and the intersects of each interaction (0-1, 0-2, 0-4, 1-2, 1-4, 2-4)

# use data frames anno_h0, anno_h1, anno_h2, anno_h4
# all DEGs is annotated_DEG_list

length(annotated_DEG_list$Row.names) #142 total

##Pairwise
h0_h1 <- intersect(anno_h0$Row.names, anno_h1$Row.names)
h0_h2 <- intersect(anno_h0$Row.names, anno_h2$Row.names)
h0_h4 <- intersect(anno_h0$Row.names, anno_h4$Row.names)
h1_h2 <- intersect(anno_h1$Row.names, anno_h2$Row.names)
h1_h4 <- intersect(anno_h1$Row.names, anno_h4$Row.names)
h2_h4 <- intersect(anno_h2$Row.names, anno_h4$Row.names)

##Triple
h0_h1_h2 <- intersect(h0_h1, anno_h2$Row.names)
h0_h2_h4 <- intersect(h0_h2, anno_h4$Row.names)
h1_h2_h4 <- intersect(h1_h2, anno_h4$Row.names)
h0_h1_h4 <- intersect(h0_h1, anno_h4$Row.names)
##Quadruple
h0_h1_h2_h4 <- intersect(h0_h1, h2_h4)

DGE_hour_venn <- draw.quad.venn(area1 = length(anno_h0$Row.names),
               area2 = length(anno_h1$Row.names),
               area3 = length(anno_h2$Row.names),
               area4 = length(anno_h4$Row.names),
               n12 = length(h0_h1),
               n13 = length(h0_h2),
               n14 = length(h0_h4),
               n23 = length(h1_h2),
               n24 = length(h1_h4),
               n34 = length(h2_h4),
               n123 = length(h0_h1_h2),
               n124 = length(h0_h1_h4),
               n134 = length(h0_h2_h4),
               n234 = length(h1_h2_h4),
               n1234 = length(h0_h1_h2_h4),
               category = c("Hour 0", "Hour 1", "Hour 2", "Hour 4"),
               fill = c("olivedrab3", "skyblue", "springgreen", "deepskyblue"),
               lwd = rep(1,4),
               cat.cex = rep(1.4, 4),
               #cat.fontfamily = rep("Arial", 4),
               cat.fontface = rep("plain", 4),
               alpha = c(0.3,0.3,0.3,0.3),
               cex = c(rep(1.4,5), 3, rep(1.4, 9)),
               fontface = c(rep("plain", 5), "bold", rep("plain", 9)),
               #fontfamily = c(rep("Arial", 15)),
               label.col = c(rep("black", 5), "red", rep("black", 9)),
               print.mode = "raw",
               sigdigs = 2,
               scaled = TRUE
) 
#pdf("~/OneDrive - University of Miami/Cnidimmunity Lab/Wound_Healing_Project/2022 updates with Sami Beasley/manuscript figures and tables/VennDiagram_DGEallhours")
grid.draw(DGE_hour_venn)
```





